src/toolbox/prover.rs

```diff
@@ -172,90 +179,204 @@ impl<'a> IsSigmaProtocol for Prover<'a> {
+                            };
+                        }
                     };
+                };
 
-                    let challenge = Scalar::random(&mut transcript_rng);
-                    shares.push(Some(challenge));
-                    RistrettoPoint::multiscalar_mul(
-                        rhs_lin_combo.iter()
-                            .map(|(_sc_var, _pt_var)| {
-                                let response = Scalar::random(&mut transcript_rng);
-                                fake_responses.push(Some(response));
-                                response
-                            })
-                            .chain(iter::once(-&challenge)),
-                        rhs_lin_combo.iter()
-                            .map(|(_sc_var, pt_var)| self.points[pt_var.0])
-                            .chain(iter::once(self.points[lhs_var.0])),
-                    )
+                // Now that we know all the clauses, make sure at least one is full, and generate one challenge share
+                // per clause (NOT per constraint!)
+                let mut have_one = false;
+                for (i, clause) in clause_tracker.iter().enumerate() {
+                    match clause {
+                        Some((terms, filled_in)) => {
+                            debug!("Clause {} has {} terms and we know {} keys", i, terms, filled_in);
+                            if terms == filled_in {
+                                // this is a fully-known clause, so we'll be able to calculate the challenge later on
+                                have_one = true;
+                                shares.push(None);
+                            } else {
+                                // this is a faked clause, so we need to randomize a challenge for it
+                                shares.push(Some(Scalar::random(&mut transcript_rng)));
+                            };
+                        },
+                        None => (),
+                    }
+                }
+                if !have_one {
+                    return Err(ProofError::InsufficientKeys)
                 }
-            };
-
-            let encoding = self
-                .transcript
-                .append_blinding_commitment(self.point_labels[lhs_var.0], &commitment);
-
-            commitments.push(encoding);
-        }
 
-        // Now that we know all the clauses, make sure at least one is full
-        let mut have_one = false;
-        for (i, clause) in clause_tracker.iter().enumerate() {
-            match clause {
-                Some((terms, filled_in)) => {
-                    debug!("Clause {} has {} terms and we know {} keys", i, terms, filled_in);
-                    if terms == filled_in { have_one = true; break; }
-                },
-                None => (),
+                // Using the shares, calculate our commitments
+                for (clause_nr, lhs_var, rhs_lin_combo) in &self.constraints {
```

Comment:
Can't point exactly to it, but it seems we have quite some duplicate code here. Can we avoid that?

Reply:
The algorithms are similar, but not identical.  We can extract some of the common sections, at the cost of runtime efficiency.  For example, in the Shamir case we calculate both the clause tracker and shares in the same loop over `self.constraints`.  We also need the clause tracker data for Xor, but running it before the `match` block adds an extra loop over `self.constraints` in the Shamir case.